### ISSUE TITLE
Duplicate / failed payments when payment processing gaps occur

### ISSUE THREAD

**Issue by emilyjeanrogers** [openfoodfoundation/openfoodnetwork#3925]
## Description

Inconsistencies have been identified with the way OFN responds when payment gaps occur (ie. payments are created and then rolled back / disappear). The following scenarios have been recorded:
a) Customer places an order and is charged twice by Stripe
b) Customer doesn't complete an order but is charged anyway
c) Customer's payment failed but customer was not notified (so order was presumed successful)

## Expected Behaviour

Processed payments should not disappear from the system. 
Errors after the payment processing have to be handled gracefully.

## Actual Behaviour

Gaps in processed payments have led to duplicate and/or incorrect charging in some cases.

Further information from @mkllnk investigation:
In Aus production there have been 1021 gaps the in the payment ids of which 9 correspond to payments that were processed and then rolled back (disappeared). 5 are failed payments and 4 are successful payments.  Of the successful payments, one is from November 2018 and the other three occurred in the last 4 weeks (May/June 2019) for the recently created Enterprise North Arm Farms. 

## Steps to Reproduce

The issue cannot be reproduced.

## Animated Gif/Screenshot

![North Arm Farms Order Confirmation $9](https://user-images.githubusercontent.com/47620442/59395473-7ad07180-8dc7-11e9-9bc0-157543a677c0.png)
![Stripe confirmation of duplicate charge $9](https://user-images.githubusercontent.com/47620442/59395475-7dcb6200-8dc7-11e9-99f0-bda6d657b6cd.png)

## Context

Customers have complained to the Enterprise (North Arm Farms) about duplicate charging, failed orders and being charged for orders that were not received. The Enterprise has had to  process refunds through Stripe. The Enterprise has also had to spend time investigating the problem. After 4 weeks of consistent issues, they now manually cross check all orders against Stripe reports to identify duplicates and issues before customers can complain.

## Severity

bug-s2: a critical feature is broken for some users: duplicated and failed payments. Could affect any user. We can't find a satisfactory workaround.

## Your Environment

* Version used: 1.31.0 
* OFN Platform instance: Aus Production

## Possible Fix

Maybe related to this past issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/3924

**Comment by kirstenalarsen** [comment]
@mkllnk  apologies I think I'm having trouble following the related issues. This is an S3 because of the logging introduced which should tell us more about what's happening, is that right? And because it might tell us that it's fixed in V2 and we just have to wait? I'm wondering because the user is being really nice about it but it's a PITA and I don't want to take our eyes off it :)

**Comment by mkllnk** [comment]
Yes. I'll keep it "In Dev" which means that I will check it regularly. This weekend we will see if my little fix (workaround) works and what other data it will give us. No active development at the moment.

Emily and I have been in contact with the affected enterprise.

**Comment by kirstenalarsen** [comment]
We have replicated this issue in V1, both for the affected user and in another enterprise (so not specific to this user data).

To replicate:
- create a product with 1 in stock
- connect stripe to shop
- put in order cycle and open shop
- have multiple users place the scarce item in the cart at the same time (system allows this)
- multiple users go to checkout at same time
- multiple users set up credit card payment through stripe
- multiple users hit 'Place Order Now' at the same time

What happens:
- one user is successful, order completes, payment is made and user is returned to the order confirmation screen
- other users (in our case two): checkout screen . . no error message, place order now button still active etc. BUT the payment has been taken in Stripe. There is no valid order in OFN

3 Stripe payments made
![image](https://user-images.githubusercontent.com/2911176/60635090-c5885b00-9e54-11e9-8ab9-f637d7c1566d.png)

One successful order created in OFN
![image](https://user-images.githubusercontent.com/2911176/60635193-392a6800-9e55-11e9-8622-22d3d39fdb97.png)

A trace of the two other customers can be found by searching under their name (having unticked 'show only completed orders'
![image](https://user-images.githubusercontent.com/2911176/60635260-81498a80-9e55-11e9-9398-613f2386a345.png)
![image](https://user-images.githubusercontent.com/2911176/60635275-8dcde300-9e55-11e9-8955-c7e75d546a15.png)

Opening these orders enables you to see what items were in the cart, but processing the payment to make them complete will exceed the stock levels set by the Hub

If the User that is still sitting on Checkout page now hits 'Place Order Now' they will be taken to the Edit Cart page with a notice that one of their items is out of stock. They can remove that item and then place the order. However, they have now paid twice. This appears as:
- customer has paid twice - 2 payments in Stripe (possibly / likely different prices)
- only one order in OFN
- confusion and angst :)

DIAGNOSIS NOTES (from discussion):
- This possibly hasn't turned up before because of the 'package' of activities undertaken in a Stripe payment - there is a much longer time between when we check stock level before processing the payment and when we check again to finalise the order
- However it also may have been happening without people really realising / noticing :(

This is definitely back to S2 - taking payments for orders that then disappear without anyone knowing about it is BAD. It could happen to any user if two customers try to buy and item going out of stock at the same time. More likely if they are using Stripe.

We are going to see if we can still replicate this in V2. Hopefully not. But if we can, there will be a strong case for upgrading this to a S2 severity.

**Comment by kirstenalarsen** [comment]
NB. This does not resolve Case c - @mkllnk should we perhaps open a separate bug report for that? Not sure of frequency / severity compared to a) and b) which I'm pretty confident are both this?

**Comment by mkllnk** [comment]
I haven't observed any more cases c. Let's solve this first and then see if there are still problems.

**Comment by mkllnk** [comment]
I continued the testing on our staging server. I could replicate the issue with v1 code. After the upgrade, the problem was solved in this way:

- Two people have an item with limited stock in their cart.
- Both check out at the same time.
- Both orders are paid (via Stripe).
- The stock levels are drawn and the product now has negative stock.
- Both orders are successful.

This behaviour is not ideal. The order confirmation should display that there is a stock problem with one of the items. And the admin ui should flag these issues. But it's a lot better and I think it's manageable for now.
